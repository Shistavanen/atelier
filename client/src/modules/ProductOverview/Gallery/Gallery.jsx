import React, { useState, useEffect } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
// import '../fontawesome.js'

export default function Gallery(props) {

  const [galleryPhoto, setGalleryPhoto] = useState(props.style.photos[0].url)
  const [imageHeight, setImageHeight] = useState(null)
  const [maxThumbnails, setMaxThumbnails] = useState(0)
  // const [highlighted, setHighlighted] = useState(false)
  const [clicked, setClicked] = useState(null)

  let [lowIndex, setLowIndex] = useState(0)
  let [highIndex, setHighIndex] = useState(7)

  //let [border, setBorder] = useState('1px solid black')

  //let portrait = false;
  const [clientHeight, setClientHeight] = useState(null)
  const [clientWidth, setClientWidth] = useState(null)

  // let styling = {cursor: 'pointer', objectFit: 'cover'};

  function getImgSize(imgSrc, callback) {
    const newImg = new Image();

    newImg.onload = function() {
      const height = newImg.height;
      const width = newImg.width;
      callback({ width, height })
    }

    newImg.src = imgSrc;
  }

  function getImgDimensions({ width, height }) {
    console.log('width height', width, height)
    if(height > width) {
      setClientHeight('600')
      setClientWidth('700')
    } else {
      setClientHeight('600')
      setClientWidth('700')
      //console.log('PORTRAIT: ', portrait)
    }
  }

  getImgSize(galleryPhoto, getImgDimensions)

  function handleThumbnailClick(i, photo) {

    // return function handleClick(photo) {
    //   //let img = document.getElementById(`galleryThumbnail${i}`)
    //   //let img = document.getElementById('galleryMainImage')
    //   //console.log('url: ', photo)
    //   //console.log('NATURAL HEIGHT: ',img.naturalHeight)
    //   //console.log('NATURAL WIDTH: ', img.naturalWidth)
    //   // styling = {cursor: 'pointer', objectFit: 'cover', border: '15px solid black'}

    //   //let imageId = `galleryThumbnail${i}`
    //   //document.getElementById(imageId).style.border = "3px solid black";
    //   //setBorder('3px solid black')

    //   setGalleryPhoto(photo.url)
    //   setClicked(clicked === i ? null : i)
    // }

    setGalleryPhoto(photo.url)
    setClicked(clicked === i ? null : i)

  }

  function mainImage(e, photo) {
    e.stopPropagation()
    setGalleryPhoto(photo)
  }

  useEffect(() => {
    setGalleryPhoto(props.style.photos[0].url)
  },[props.style])

  // let img;
  // let height;

  // useEffect(() => {
  //   img = document.getElementById('galleryMainImage')
  //   height = img.clientHeight
  //   setImageHeight(height)
  //   console.log('HEIGHT',height)
  //   console.log('WIDTH', img.clientWidth)
  // }, [])

  // useEffect(() => {
  //   img = document.getElementById('galleryMainImage')
  //   height = img.clientHeight
  //   setImageHeight(height)
  //   console.log('HEIGHT',height)
  //   console.log('WIDTH', img.clientWidth)
  // }, [galleryPhoto])

  // useEffect(()=> { //TODO: add condition for 6 thumbnails
  //   if(clientHeight === '467') {
  //     setMaxThumbnails(5)
  //   } else {
  //     setMaxThumbnails(7)
  //   }
  //   console.log('MAX THUMBNAILS: ', maxThumbnails)
  // }, [clientHeight])
  //width='300' height='200'

  function handleDownArrow() {
    setLowIndex(lowIndex += 1)
    setHighIndex(highIndex += 1)
  }

  function handleUpArrow() {
    setLowIndex(lowIndex -= 1)
    setHighIndex(highIndex -= 1)
  }

  function handleRightArrow() {
    let nextThumbnail = document.getElementById(`galleryThumbnail${i + 1}`)
    setGalleryPhoto(nextThumbnail.url)
  }


  return (
    <div id='gallery' data-testid="galleryTest">
      <img id='galleryMainImage' src={galleryPhoto} width={clientWidth} height={clientHeight}  />
      <FontAwesomeIcon icon='fa-solid fa-circle-chevron-left' id='galleryLeft' cursor={'pointer'} />
      <FontAwesomeIcon icon='fa-solid fa-circle-chevron-right' id='galleryRight' cursor={'pointer'} />
      {/* key={props.style.style_id}> */}
      <br></br>

      <div className='galleryThumbnailContainer'>
        {lowIndex === 0 ? <div><br/></div> : <FontAwesomeIcon icon='fa-solid fa-chevron-up' onClick={handleUpArrow} cursor={'pointer'}/> }

      {/* <i class='fa-solid fa-angle-up'></i> */}
        {props.style.photos.map((photo, i) => {

          if(i >= lowIndex && i < highIndex) {
            return (
              <div onClick={()=>handleThumbnailClick(i, photo)} key={i}>
                {clicked === i ?
                  <img className='galleryThumbnail'
                    id={`galleryThumbnail${i}`}
                    //key={i}
                    src={photo.thumbnail_url}
                    //onClick={()=>mainImage(photo)}
                    style={{cursor: 'pointer', objectFit: 'cover', boxShadow: '3px 3px'}}
                    width='60' height='60'
                  /> :
                  <img className='galleryThumbnail'
                    id={`galleryThumbnail${i}`}
                    //key={i}
                    src={photo.thumbnail_url}
                    //onClick={()=>click(i)(photo)}
                    style={{cursor: 'pointer', objectFit: 'cover', border: '1px solid black'}}
                    width='60' height='60'
                  />
                }

              </div>
            )

          }
        })}
        {highIndex === props.style.photos.length - 1 ? <></> : <FontAwesomeIcon icon='fa-solid fa-chevron-down' onClick={handleDownArrow} cursor={'pointer'}/> }

      </div>
    </div>
  )

}