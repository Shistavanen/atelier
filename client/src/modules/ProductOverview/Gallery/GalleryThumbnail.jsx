import React, { useState } from 'react'

export default function GalleryThumbnail(props) {


  return (

  )
}