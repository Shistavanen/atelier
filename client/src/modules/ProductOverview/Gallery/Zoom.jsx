import React, { useState, useEffect } from 'react'


export default function Zoom(props) {

  let width = '1500px';
  let height = '800px';

  return (
    <div id='zoomView'
         onClick={props.expandView}
         style={{
           cursor: 'zoom-out',
           backgroundRepeat: 'no-repeat',
           backgroundSize: `${props.naturalWidth * 2.5}px ${props.naturalHeight * 2.5}px`,
           backgroundImage: `url(${props.galleryPhoto})`,
           height: 'auto',
           minHeight: height,
           minWidth: width,
           overflow: 'hidden',
           backgroundPosition: 'center',
           position: 'absolute',
           width: '100%',
           zIndex: '10'
         }}
         >
    </div>
  )
}