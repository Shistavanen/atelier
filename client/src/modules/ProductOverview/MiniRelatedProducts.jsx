import React, { useState, useEffect} from 'react'
import axios from 'axios'

export default function MiniRelatedProducts(props) {

  function handleClick(productId) {
    console.log('CLICKEDDDD')
    props.setProductId(productId)

    axios.get(`/relatedProducts/${productId}`)
      .then(results => {
        console.log('PRODUCT THAT IS RELATED: ', results.data)
        props.setProduct(results.data)
      })
      .catch(err => console.log(err))
  }

  return (
    <div> -Related Products-
      <ul>
        {props.relatedProducts.map(product => (
          <li onClick={()=>handleClick(product)} style={{cursor: 'pointer'}} key={product}>{product}</li>
        ))}
      </ul>
    </div>
  )
}