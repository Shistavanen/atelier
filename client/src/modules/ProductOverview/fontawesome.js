import { library } from '@fortawesome/fontawesome-svg-core';

import {
  faAngleUp,
  faAngleDown,
  faChevronUp,
  faChevronDown,
  faChevronLeft,
  faChevronRight
} from '@fortawesome/free-solid-svg-icons';

library.add(
  faAngleUp,
  faAngleDown,
  faChevronUp,
  faChevronDown,
  faChevronLeft,
  faChevronRight
)