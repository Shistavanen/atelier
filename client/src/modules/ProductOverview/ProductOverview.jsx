import React, { useState, useEffect } from 'react'
import './fontawesome.js'
import axios from 'axios'
import Category from './ProductInfo/Category.jsx'
import Title from './ProductInfo/Title.jsx'
import Price from './ProductInfo/Price.jsx'
import StarRating from './ProductInfo/StarRating.jsx'
import Overview from './ProductInfo/Overview.jsx'
import MyOutfitButton from './ProductInfo/MyOutfitButton.jsx'
import StyleSelector from './StyleSelector/StyleSelector.jsx'
import Cart from './Cart/Cart.jsx'

import Gallery from './Gallery/Gallery.jsx'
import MiniRelatedProducts from './MiniRelatedProducts.jsx'

export default function ProductOverview({ product, productId, setProductId, setProduct }) {

  function getCurrentProductData(id) {
    return axios.get(`/productOverview/${id}`) //orchestrates endpoint
      .then(results => {
        return results.data
    })
  }

  const [currentProduct, setCurrentProduct] = useState(product)
  const [category, setCategory] = useState(product.category)
  const [title, setTitle] = useState(product.name)
  const [price, setPrice] = useState(product.default_price)
  const [slogan, setSlogan] = useState(product.slogan)
  const [description, setDescription] = useState(product.description)
  const [id, setId] = useState(productId)

  const [features, setFeatures] = useState([])
  const [reviews, setReviews] = useState([])

  const [styles, setStyles] = useState([])
  const [style, setStyle] = useState(null)
  const [salePrice, setSalePrice] = useState(null)

  const [hideInfo, setHideInfo] = useState(false)

  const [styleTitle, setStyleTitle] = useState('')
  const [highlightedThumbnail, setHighlightedThumbnail] = useState(0)

  const [relatedProducts, setRelatedProducts] = useState([])


  useEffect(()=> {

    getCurrentProductData(id)
      .then(data => {
        console.log("STYLES FROM GET REQ: ", data)
        setFeatures(data.features)
        setReviews(Object.entries(data.reviews))
        setRelatedProducts(data.related)
        setStyles(data.styles)
        setStyle(data.styles[0])
        setStyleTitle(data.styles[0].name)
        setSalePrice(data.styles[0].sale_price)
      })

  }, [])

  //CHECK RELATED PRODUCTS
  useEffect(() => {
    console.log('RELATED PRODUCTS',relatedProducts)
  }, [relatedProducts])

  if(!style || !relatedProducts) return null

  return (
    <div data-testid="productOverviewTest" style={{marginTop: '25px', marginBottom: '30px'}}> {/* increases surrounding margins to approx centering */}
      <div id='upper' style={{display: 'flex'}}>
        <Gallery style={style} setHideInfo={setHideInfo} />
        {hideInfo ? null :
          <div style={{marginLeft: '60px', color: '#25383C'}}>
            <StarRating reviews={reviews}/>
            <br/>
            <Category category={category} />

            <Title title={title} />
            <br />
            <Price price={price} salePrice={salePrice} />
            <br />

            <StyleSelector
              styles={styles}
              style={style}
              setStyle={setStyle}
              styleTitle={styleTitle}
              setStyleTitle={setStyleTitle}
              setPrice={setPrice}
              setSalePrice={setSalePrice}
              highlightedThumbnail={highlightedThumbnail}
              setHighlightedThumbnail={setHighlightedThumbnail}/>
            <br/>

            <Cart style={style} setStyle={setStyle} />
          </div>
         }
        <br></br>

      </div>
      <Overview slogan={slogan} description={description} features={features} />
      <MiniRelatedProducts relatedProducts={relatedProducts} setProductId={setProductId} setProduct={setProduct}/>
    </div>
  )

}