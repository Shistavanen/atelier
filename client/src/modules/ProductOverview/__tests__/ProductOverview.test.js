/**
 * @jest-environment jsdom
 */

import ReactDOM from 'react-dom'
import React from 'react';
import '@testing-library/jest-dom'
import {render, fireEvent, screen, cleanup} from '@testing-library/react'
import axios from 'axios'
import regeneratorRuntime from "regenerator-runtime";
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import Gallery from '../Gallery/Gallery.jsx'
import ProductOverview from '../ProductOverview.jsx'

// it('Renders ProductOverview component', () => {
//   axios.get('/products', {headers: {'Authorization': APIKEY}})
//     .then(products => {
//       return products.data[4]
//     })
//     .then(product => {
//       const component = renderer.create(<ProductOverview product={product} productId={product.id}/>)
//       let tree = component.toJSON()
//       expect(tree).toMatchSnapshot()

//     })
// })

const server = setupServer(
  rest.get('/productOverview/:product_id', (req, res, ctx) => {
    return res(
      ctx.json({
        features: [{feature: 'Sole', value: 'Rubber'}, {feature: 'Material', value: 'FullControlSkin'}],
        reviews: {1: '2', 2: '14', 3: '2', 4: '8', 5: '3'},
        styles: [{
          "default?" : true,
          name: 'White & White',
          original_price: '99.00',
          photos: [{
            "thumbnail_url": "https://images.unsplash.com/photo-1544441892-794166f1e3be?ixlib=rb-1.2.1&auto=format&fit=crop&w=300&q=80",
            "url": "https://images.unsplash.com/photo-1544441892-794166f1e3be?ixlib=rb-1.2.1&auto=format&fit=crop&w=1650&q=80"
          }, {
            "thumbnail_url": "https://images.unsplash.com/photo-1514590734052-344a18719611?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
            "url": "https://images.unsplash.com/photo-1514590734052-344a18719611?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1650&q=80"
          }],
          sale_price: null,
          skus: {
            "2580652": {
              "quantity": 14,
              "size": "7"
            },
            "2580653": {
              "quantity": 25,
              "size": "7.5"
            },
            "2580654": {
              "quantity": 9,
              "size": "8"
            },
            "2580655": {
              "quantity": 2,
              "size": "8.5"
            },
            "2580656": {
              "quantity": 18,
              "size": "9"
            },
            "2580657": {
              "quantity": 12,
              "size": "9.5"
            },
            "2580658": {
              "quantity": 10,
              "size": "10"
            },
            "2580659": {
              "quantity": 18,
              "size": "10.5"
            },
            "2580660": {
              "quantity": 11,
              "size": "11"
            },
            "2580661": {
              "quantity": 35,
              "size": "11.5"
            },
            "2580662": {
              "quantity": 25,
              "size": "12"
            }
          },
          style_id: 444243
        },
        {
          "style_id": 444244,
          "name": "White & Red",
          "original_price": "99.00",
          "sale_price": null,
          "default?": false,
          "photos": [
            {
              "thumbnail_url": "https://images.unsplash.com/photo-1542280756-74b2f55e73ab?ixlib=rb-1.2.1&auto=format&fit=crop&w=300&q=80",
              "url": "https://images.unsplash.com/photo-1542280756-74b2f55e73ab?ixlib=rb-1.2.1&auto=format&fit=crop&w=1650&q=80"
            },
            {
              "thumbnail_url": "https://images.unsplash.com/photo-1521093470119-a3acdc43374a?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
              "url": "https://images.unsplash.com/photo-1521093470119-a3acdc43374a?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80"
            },
            {
              "thumbnail_url": "https://images.unsplash.com/photo-1474494819794-90f9664b530d?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
              "url": "https://images.unsplash.com/photo-1474494819794-90f9664b530d?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1652&q=80"
            },
            {
              "thumbnail_url": "https://images.unsplash.com/photo-1465877783223-4eba513e27c6?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
              "url": "https://images.unsplash.com/photo-1465877783223-4eba513e27c6?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1650&q=80"
            },
            {
              "thumbnail_url": "https://images.unsplash.com/photo-1428790031246-698d71b6fe3f?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
              "url": "https://images.unsplash.com/photo-1428790031246-698d71b6fe3f?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1650&q=80"
            },
            {
              "thumbnail_url": "https://images.unsplash.com/photo-1470282312847-28b943046dc1?ixlib=rb-1.2.1&auto=format&fit=crop&w=300&q=80",
              "url": "https://images.unsplash.com/photo-1470282312847-28b943046dc1?ixlib=rb-1.2.1&auto=format&fit=crop&w=1652&q=80"
            },
            {
              "thumbnail_url": "https://images.unsplash.com/photo-1428790067070-0ebf4418d9d8?ixlib=rb-1.2.1&auto=format&fit=crop&w=300&q=80",
              "url": "https://images.unsplash.com/photo-1428790067070-0ebf4418d9d8?ixlib=rb-1.2.1&auto=format&fit=crop&w=1650&q=80"
            },
            {
              "thumbnail_url": "https://images.unsplash.com/photo-1451256656121-9ffc0c898a49?ixlib=rb-1.2.1&auto=format&fit=crop&w=300&q=80",
              "url": "https://images.unsplash.com/photo-1451256656121-9ffc0c898a49?ixlib=rb-1.2.1&auto=format&fit=crop&w=1650&q=80"
            },
            {
              "thumbnail_url": "https://images.unsplash.com/photo-1524604519054-2365ac11e431?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
              "url": "https://images.unsplash.com/photo-1524604519054-2365ac11e431?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=975&q=80"
            },
            {
              "thumbnail_url": "https://images.unsplash.com/photo-1541006008768-d181e7f9f3d9?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
              "url": "https://images.unsplash.com/photo-1541006008768-d181e7f9f3d9?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1568&q=80"
            },
            {
              "thumbnail_url": "https://images.unsplash.com/photo-1430390456011-25ac9244999c?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
              "url": "https://images.unsplash.com/photo-1430390456011-25ac9244999c?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80"
            }
          ],
          "skus": {
            "2580663": {
              "quantity": 14,
              "size": "7"
            },
            "2580664": {
              "quantity": 25,
              "size": "7.5"
            },
            "2580665": {
              "quantity": 9,
              "size": "8"
            },
            "2580666": {
              "quantity": 2,
              "size": "8.5"
            },
            "2580667": {
              "quantity": 18,
              "size": "9"
            },
            "2580668": {
              "quantity": 12,
              "size": "9.5"
            },
            "2580669": {
              "quantity": 10,
              "size": "10"
            },
            "2580670": {
              "quantity": 18,
              "size": "10.5"
            },
            "2580671": {
              "quantity": 11,
              "size": "11"
            },
            "2580672": {
              "quantity": 35,
              "size": "11.5"
            },
            "2580673": {
              "quantity": 25,
              "size": "12"
            }
          }
        }
      ]
      })
    )
  })
)

beforeAll(() => {
  server.listen()
})

afterAll(() => {
  server.close()
})

describe.only('Product Overview', () => {
  it('should render Gallery Component', () => {
    render(<Gallery style={{
      "default?" : true,
      name: 'White & White',
      original_price: '99.00',
      photos: [{
        "thumbnail_url": "https://images.unsplash.com/photo-1544441892-794166f1e3be?ixlib=rb-1.2.1&auto=format&fit=crop&w=300&q=80",
        "url": "https://images.unsplash.com/photo-1544441892-794166f1e3be?ixlib=rb-1.2.1&auto=format&fit=crop&w=1650&q=80"
      }, {
        "thumbnail_url": "https://images.unsplash.com/photo-1514590734052-344a18719611?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
        "url": "https://images.unsplash.com/photo-1514590734052-344a18719611?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1650&q=80"
      }],
      sale_price: null,
      skus: {
        "2580652": {
          "quantity": 14,
          "size": "7"
        },
        "2580653": {
          "quantity": 25,
          "size": "7.5"
        },
        "2580654": {
          "quantity": 9,
          "size": "8"
        },
        "2580655": {
          "quantity": 2,
          "size": "8.5"
        },
        "2580656": {
          "quantity": 18,
          "size": "9"
        },
        "2580657": {
          "quantity": 12,
          "size": "9.5"
        },
        "2580658": {
          "quantity": 10,
          "size": "10"
        },
        "2580659": {
          "quantity": 18,
          "size": "10.5"
        },
        "2580660": {
          "quantity": 11,
          "size": "11"
        },
        "2580661": {
          "quantity": 35,
          "size": "11.5"
        },
        "2580662": {
          "quantity": 25,
          "size": "12"
        }
      },
      style_id: 444243
    }} />)
    expect(screen.getByTestId('galleryTest')).toBeInTheDocument();
  })
})