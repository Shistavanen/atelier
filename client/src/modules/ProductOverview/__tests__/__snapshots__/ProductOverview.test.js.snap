// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders ProductOverview component 1`] = `<div />`;
